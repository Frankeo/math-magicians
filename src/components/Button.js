import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class Button extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  render() {
    const { name, className, handleClick } = this.props;
    return (
      <button type="button" className={className} id={name} name={name} onClick={handleClick}>{name}</button>
    );
  }
}

Button.propTypes = {
  handleClick: PropTypes.func.isRequired,
  name: PropTypes.string.isRequired,
  className: PropTypes.string.isRequired,
};
