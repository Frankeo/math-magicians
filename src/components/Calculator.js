import React, { Component } from 'react';
import calculate from '../logic/calculate';

class Calculator extends Component {
  constructor(props) {
    super(props);
    this.state = {
      total: 0,
      next: null,
      operation: null,
    };

    this.handleClick = this.handleClick.bind(this);
  }

  handleClick(e) {
    e.preventDefault();
    this.setState((prev) => calculate(prev, e.target.name));
  }

  render() {
    const { total, next, operation } = this.state;
    const buttons = [['AC', '+/-', '%', 'รท'], ['7', '8', '9', 'x'], ['4', '5', '6', '-'], ['1', '2', '3', '+']].map((row) => (
      <div className="row" key={row[0]}>
        <button type="button" className="col" id={row[0]} name={row[0]} onClick={this.handleClick}>{row[0]}</button>
        <button type="button" className="col" id={row[1]} name={row[1]} onClick={this.handleClick}>{row[1]}</button>
        <button type="button" className="col" id={row[2]} name={row[2]} onClick={this.handleClick}>{row[2]}</button>
        <button type="button" className="col operature" id={row[3]} name={row[3]} onClick={this.handleClick}>{row[3]}</button>
      </div>
    ));
    return (
      <>
        <div className="calc-container">
          <form id="form">
            <input type="text" id="result" value={(total || 0) + (operation || '') + (next || '')} disabled />
          </form>
          <div className="container">
            {buttons}
            <div className="row">
              <button type="button" className="col-6" name="0" onClick={this.handleClick}>0</button>
              <button type="button" className="col-3" name="." onClick={this.handleClick}>.</button>
              <button type="button" className="col-3 operature" name="=" onClick={this.handleClick}>=</button>
            </div>
          </div>
        </div>
      </>

    );
  }
}

export default Calculator;
