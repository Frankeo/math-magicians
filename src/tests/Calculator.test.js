/* eslint-disable no-restricted-syntax */
import React from 'react';
import {
  render, screen, cleanup, fireEvent,
} from '@testing-library/react';
import Button from '../components/Button';
import Calculator from '../components/Calculator';

describe('calculator testing', () => {
  test('2 x 3', () => {
    render(<Calculator />);

    fireEvent.click(screen.getByText('2'));
    fireEvent.click(screen.getByText('x'));
    fireEvent.click(screen.getByText('3'));
    fireEvent.click(screen.getByText('='));

    expect(screen.getByText('6')).toMatchSnapshot();
  });

  test('calculator contains all buttons', () => {
    const calcButtons = ['AC', '+/-', '!', 'รท', '7', '8', '9', 'x', '4', '5', '6', '-', '1', '2', '3', '+', '0', '.', '='];
    const handleClick = (e) => {
      expect(e.target.name).toEqual('รท');
    };

    render(
      <Calculator>
        <Button name={calcButtons[3]} id={calcButtons[3]} className="col-3 operature" handleClick={handleClick} />
      </Calculator>,
    );

    for (const btn of calcButtons) {
      expect(screen.getByText(btn)).toBeInTheDocument();
    }
  });

  test('clicking the button toggles the number', () => {
    render(<Calculator />);
    const button = screen.getByTestId('1');
    fireEvent.click(button);
    expect(button).toBeInTheDocument();
  });
});

afterEach(cleanup);
