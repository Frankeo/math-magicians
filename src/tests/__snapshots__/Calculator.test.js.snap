// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculator testing 2 x 3 1`] = `
<button
  class="col-3"
  data-testid="6"
  id="6"
  name="6"
  type="button"
>
  6
</button>
`;
